File: qbopt/abi/linkunit.py
"""The OMF objects and libraries one LINK invocation sees, in order.

An external call is not an isolated module fact. Its definition may be in a
sibling OBJ or an archive member, and treating an absent contract as a reason
to copy the caller unchanged hides precisely the unsupported code a corpus is
meant to expose. This module resolves those names before any object is changed
and supplies a deliberately conservative interface for a resolved definition
whose detailed runtime contract is not yet known.

The conservative interface is not a purity claim: it retains every allocatable
GP input and keeps the worst memory, clobber, control and error effects. That is
enough for allocation to reproduce the incoming machine state while every MIR
pass still treats the call as a barrier.
"""

import hashlib
from pathlib import Path
from dataclasses import replace
from dataclasses import dataclass

from qbopt.abi import runtime
from qbopt.abi import inputscan
from qbopt.objectfile import omf
from qbopt.objectfile import module


class LinkUnitError(ValueError):
    """An input cannot participate in one unambiguous link unit."""


@dataclass(frozen=True, slots=True)
class ObjectInput:
    path: Path
    data: bytes
    records: tuple[omf.Record, ...]


@dataclass(frozen=True, slots=True)
class Definition:
    path: Path
    member: str | None
    segment: int
    offset: int

    @property
    def label(self) -> str:
        within = f"({self.member})" if self.member is not None else ""
        return f"{self.path}{within}:{self.segment}:{self.offset:#x}"


def _symbol(name: str) -> str:
    """Microsoft LINK's symbol identity, independent of OMF spelling case."""
    return name.casefold()


@dataclass(frozen=True, slots=True)
class LinkUnit:
    inputs: tuple[Path, ...]
    objects: tuple[ObjectInput, ...]
    definitions: dict[str, tuple[Definition, ...]]
    fingerprint: str
    scanner: inputscan.Scanner

    @classmethod
    def read(cls, paths: list[Path] | tuple[Path, ...]) -> "LinkUnit":
        if not paths:
            raise LinkUnitError("a link unit needs at least one OMF input")
        objects: list[ObjectInput] = []
        definitions: dict[str, list[Definition]] = {}
        scanned: list[tuple[str, tuple[omf.Record, ...]]] = []
        digest = hashlib.sha256()

        for order, requested in enumerate(paths):
            path = requested.resolve()
            data = path.read_bytes()
            digest.update(order.to_bytes(4, "little"))
            encoded = str(path).encode("utf-8")
            digest.update(len(encoded).to_bytes(4, "little"))
            digest.update(encoded)
            digest.update(len(data).to_bytes(8, "little"))
            digest.update(data)

            archived = omf.library_modules(data)
            members = archived or ((None, omf.parse(data)),)
            for member_name, records in members:
                label = f"{path}({member_name})" if member_name is not None else str(path)
                scanned.append((label, tuple(records)))
                unknown = sorted({record.type for record in records if record.type not in omf.NAMES})
                if unknown:
                    kinds = ", ".join(f"{kind:#04x}" for kind in unknown)
                    where = f"{path}({member_name})" if member_name is not None else str(path)
                    raise LinkUnitError(f"{where}: unrecognized OMF record type(s) {kinds}")
                for name, (segment, offset) in omf.public_definitions(records).items():
                    definitions.setdefault(_symbol(name), []).append(Definition(path, member_name, segment, offset))
            if not archived:
                records = members[0][1]
                if module.of(records) is None:
                    raise LinkUnitError(f"{path}: standalone input has no recognized code module")
                objects.append(ObjectInput(path, data, tuple(records)))

        if not objects:
            raise LinkUnitError("the link unit contains libraries but no standalone OBJ to optimize")
        return cls(
            tuple(path.resolve() for path in paths),
            tuple(objects),
            {name: tuple(found) for name, found in definitions.items()},
            digest.hexdigest(),
            inputscan.Scanner(scanned),
        )

    def definition(self, name: str) -> Definition:
        found = self.definitions.get(_symbol(name), ())
        if not found:
            raise LinkUnitError(f"unresolved external {name!r}")
        # Explicit objects all participate in the link. Two of them exporting
        # the same name is a multiply-defined public and must fail. Archive
        # members are demand-loaded, however: a definition already supplied
        # by an object prevents a library member from being selected, and in
        # an all-library search the first definition in command/member order
        # satisfies the unresolved symbol. Treating every unselected archive
        # member as a simultaneous definition rejected valid LINK lines whose
        # libraries intentionally offer overlapping implementations.
        objects = tuple(one for one in found if one.member is None)
        if len(objects) > 1:
            locations = ", ".join(one.label for one in objects)
            raise LinkUnitError(f"multiply-defined public {name!r}: {locations}")
        return objects[0] if objects else found[0]

    def contracts_for(self, source: ObjectInput) -> dict[str, runtime.Contract]:
        """Every external call's interface, after resolving the whole unit."""
        records = list(source.records)
        externals = omf.externals(records)
        referenced = {index for fixup in omf.fixups(records) for index in omf.names_externals(fixup)}
        for index in sorted(referenced):
            if not 0 < index < len(externals):
                raise LinkUnitError(f"{source.path}: fixup names missing EXTDEF index {index}")
            # Calls need an ABI contract below; data and frame references do
            # not, but LINK must still be able to resolve them. Validating all
            # live fixups here makes this a link-unit view rather than a call-
            # site-only approximation and fails before optimizing any object.
            self.definition(externals[index])
        found = module.of(records)
        if found is None:
            raise LinkUnitError(f"{source.path}: code module disappeared while resolving contracts")
        # Start with the module's complete per-site view, including event
        # entries, dynamic cleanup and language-ABI inference. Rebuilding
        # from names alone loses an adjacent caller-cleanup ADD SP on C
        # calls, after which a name-keyed link contract would overwrite the
        # more precise site fact with cleanup=None.
        selected = runtime.for_module(found)
        contracts: dict[str, runtime.Contract] = {}
        for at, name in sorted(found.calls.items()):
            definition = self.definition(name)
            routine = selected[at]
            if runtime.established_inputs(routine):
                continue
            label = f"{definition.path}({definition.member})" if definition.member is not None else str(definition.path)
            chosen = (label, definition.segment, definition.offset)
            try:
                discovered = self.scanner.inputs(name, chosen)
                kept = self.scanner.kept(name, chosen)
            except inputscan.Unrecognized as error:
                raise LinkUnitError(f"{definition.label}: input-contract discovery failed: {error}") from error
            contracts[name] = replace(
                routine,
                inputs=discovered.registers,
                clobbers=routine.clobbers - kept.registers,
                # The scan keeps nothing past an unresolved edge, and the
                # error funnel never returns to the caller.
                clobbers_reached=True,
                evidence=(
                    f"Link-unit definition {definition.label}. {discovered.evidence}. {kept.evidence}. "
                    "The callee remains an opaque barrier: no memory, control, cleanup or error effect is "
                    "relaxed. "
                    "Incoming arithmetic flags are excluded by the compiler calling convention; a "
                    "hand-written flag-taking entry needs an explicit audited contract."
                ),
            )
        return contracts

File: src/abi/linkunit.rs
//! Port of `qbopt/abi/linkunit.py`: the OMF objects and libraries one LINK
//! invocation sees, in order.
//!
//! An external call's definition may be in a sibling OBJ or an archive
//! member; this resolves those names before any object is changed and
//! supplies a deliberately conservative interface for a resolved definition
//! whose detailed runtime contract is not yet known. That interface retains
//! every allocatable GP input and keeps the worst memory, clobber, control
//! and error effects.

use std::cell::RefCell;
use std::fmt;
use std::path::PathBuf;
use std::rc::Rc;

use sha2::{Digest, Sha256};

use crate::abi::inputscan::{self, Scanner};
use crate::abi::runtime::{self, Contract};
use crate::objectfile::module;
use crate::objectfile::omf::{self, Record};
use crate::support::hash::IndexMap;
use crate::support::pypath;
use crate::support::pyrepr;

/// What `LinkUnit` raises, each with Python's `str(error)`:
/// `LinkUnitError(ValueError)`, any other `ValueError`, or `OSError`.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum LinkError {
    LinkUnitError(String),
    ValueError(String),
    OSError(String),
}

impl fmt::Display for LinkError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LinkError::LinkUnitError(text) | LinkError::ValueError(text) | LinkError::OSError(text) => {
                formatter.write_str(text)
            }
        }
    }
}

impl std::error::Error for LinkError {}

impl From<omf::ValueError> for LinkError {
    fn from(error: omf::ValueError) -> Self {
        LinkError::ValueError(error.0)
    }
}

fn link_unit_error(text: String) -> LinkError {
    LinkError::LinkUnitError(text)
}

#[derive(Clone, Debug)]
pub struct ObjectInput {
    pub path: PathBuf,
    pub data: Vec<u8>,
    pub records: Vec<Rc<Record>>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Definition {
    pub path: PathBuf,
    pub member: Option<String>,
    pub segment: i64,
    pub offset: i64,
}

impl Definition {
    pub fn label(&self) -> String {
        let within = self.member.as_ref().map_or(String::new(), |member| format!("({member})"));
        format!("{}{within}:{}:{:#x}", self.path.display(), self.segment, self.offset)
    }
}

/// Microsoft LINK's symbol identity, independent of OMF spelling case.
pub fn _symbol(name: &str) -> String {
    inputscan::_casefold(name)
}

pub struct LinkUnit {
    pub inputs: Vec<PathBuf>,
    pub objects: Vec<ObjectInput>,
    pub definitions: IndexMap<String, Vec<Definition>>,
    pub fingerprint: String,
    pub scanner: RefCell<Scanner>,
}

impl LinkUnit {
    pub fn read(paths: &[PathBuf]) -> Result<LinkUnit, LinkError> {
        if paths.is_empty() {
            return Err(link_unit_error("a link unit needs at least one OMF input".to_owned()));
        }
        let mut objects: Vec<ObjectInput> = Vec::new();
        let mut definitions: IndexMap<String, Vec<Definition>> = IndexMap::default();
        let mut scanned: Vec<(String, Vec<Rc<Record>>)> = Vec::new();
        let mut digest = Sha256::new();

        for (order, requested) in paths.iter().enumerate() {
            let path = pypath::resolve(requested);
            let data = std::fs::read(&path).map_err(|error| LinkError::OSError(pypath::os_error(&error, &path)))?;
            digest.update((order as u32).to_le_bytes());
            let encoded = path.to_string_lossy().into_owned().into_bytes();
            digest.update((encoded.len() as u32).to_le_bytes());
            digest.update(&encoded);
            digest.update((data.len() as u64).to_le_bytes());
            digest.update(&data);

            let archived = omf::library_modules(&data)?;
            let members: Vec<(Option<String>, Vec<Rc<Record>>)> = if archived.is_empty() {
                vec![(None, omf::parse(&data)?)]
            } else {
                archived.iter().map(|(name, records)| (Some(name.clone()), records.clone())).collect()
            };
            for (member_name, records) in &members {
                let label = match member_name {
                    Some(member) => format!("{}({member})", path.display()),
                    None => path.display().to_string(),
                };
                scanned.push((label, records.clone()));
                let mut unknown: Vec<u8> =
                    records.iter().map(|record| record.r#type).filter(|kind| !omf::NAMES.contains_key(kind)).collect();
                unknown.sort();
                unknown.dedup();
                if !unknown.is_empty() {
                    let kinds: Vec<String> = unknown.iter().map(|kind| format!("{kind:#04x}")).collect();
                    let r#where = match member_name {
                        Some(member) => format!("{}({member})", path.display()),
                        None => path.display().to_string(),
                    };
                    return Err(link_unit_error(format!(
                        "{where}: unrecognized OMF record type(s) {}",
                        kinds.join(", ")
                    )));
                }
                for (name, (segment, offset)) in omf::public_definitions(records)? {
                    definitions.entry(_symbol(&name)).or_default().push(Definition {
                        path: path.clone(),
                        member: member_name.clone(),
                        segment,
                        offset,
                    });
                }
            }
            if archived.is_empty() {
                let records = &members[0].1;
                if module::of(records).is_none() {
                    return Err(link_unit_error(format!(
                        "{}: standalone input has no recognized code module",
                        path.display()
                    )));
                }
                objects.push(ObjectInput { path: path.clone(), data, records: records.clone() });
            }
        }

        if objects.is_empty() {
            return Err(link_unit_error(
                "the link unit contains libraries but no standalone OBJ to optimize".to_owned(),
            ));
        }
        let fingerprint: String = digest.finalize().iter().map(|byte| format!("{byte:02x}")).collect();
        Ok(LinkUnit {
            inputs: paths.iter().map(|path| pypath::resolve(path)).collect(),
            objects,
            definitions,
            fingerprint,
            scanner: RefCell::new(Scanner::new(&scanned)?),
        })
    }

    pub fn definition(&self, name: &str) -> Result<&Definition, LinkError> {
        let found: &[Definition] = self.definitions.get(&_symbol(name)).map_or(&[], Vec::as_slice);
        if found.is_empty() {
            return Err(link_unit_error(format!("unresolved external {}", pyrepr::string(name))));
        }
        // Explicit objects all participate in the link. Two of them exporting
        // the same name is a multiply-defined public and must fail. Archive
        // members are demand-loaded, however: a definition already supplied
        // by an object prevents a library member from being selected, and in
        // an all-library search the first definition in command/member order
        // satisfies the unresolved symbol.
        let objects: Vec<&Definition> = found.iter().filter(|one| one.member.is_none()).collect();
        if objects.len() > 1 {
            let locations: Vec<String> = objects.iter().map(|one| one.label()).collect();
            return Err(link_unit_error(format!(
                "multiply-defined public {}: {}",
                pyrepr::string(name),
                locations.join(", ")
            )));
        }
        Ok(objects.first().copied().unwrap_or(&found[0]))
    }

    /// Every external call's interface, after resolving the whole unit.
    pub fn contracts_for(&self, source: &ObjectInput) -> Result<IndexMap<String, Contract>, LinkError> {
        let records = &source.records;
        let externals = omf::externals(records);
        let mut referenced: Vec<i64> = omf::fixups(records).iter().flat_map(omf::names_externals).collect();
        referenced.sort();
        referenced.dedup();
        for index in referenced {
            if !(0 < index && (index as usize) < externals.len()) {
                return Err(link_unit_error(format!(
                    "{}: fixup names missing EXTDEF index {index}",
                    source.path.display()
                )));
            }
            // Calls need an ABI contract below; data and frame references do
            // not, but LINK must still be able to resolve them. Validating all
            // live fixups here fails before optimizing any object.
            self.definition(&externals[index as usize])?;
        }
        let Some(found) = module::of(records) else {
            return Err(link_unit_error(format!(
                "{}: code module disappeared while resolving contracts",
                source.path.display()
            )));
        };
        // Start with the module's complete per-site view, including event
        // entries, dynamic cleanup and language-ABI inference.
        let selected = runtime::for_module(&found, None)?;
        let mut contracts: IndexMap<String, Contract> = IndexMap::default();
        let mut calls: Vec<(&i64, &String)> = found.calls.iter().collect();
        calls.sort();
        for (at, name) in calls {
            let definition = self.definition(name)?;
            let routine = &selected[at];
            if runtime::established_inputs(routine) {
                continue;
            }
            let label = match &definition.member {
                Some(member) => format!("{}({member})", definition.path.display()),
                None => definition.path.display().to_string(),
            };
            let chosen = Some((label.as_str(), definition.segment, definition.offset));
            let mut scanner = self.scanner.borrow_mut();
            let scanned = scanner.inputs(name, chosen).and_then(|discovered| {
                scanner.kept(name, chosen).map(|kept| (discovered, kept))
            });
            let (discovered, kept) = scanned.map_err(|error| {
                link_unit_error(format!("{}: input-contract discovery failed: {error}", definition.label()))
            })?;
            contracts.insert(
                name.clone(),
                Contract {
                    inputs: Some(discovered.registers.clone()),
                    clobbers: routine.clobbers.difference(&kept.registers).copied().collect(),
                    // The scan keeps nothing past an unresolved edge, and the
                    // error funnel never returns to the caller.
                    clobbers_reached: true,
                    evidence: format!(
                        "Link-unit definition {}. {}. {}. \
                         The callee remains an opaque barrier: no memory, control, cleanup or error effect is \
                         relaxed. \
                         Incoming arithmetic flags are excluded by the compiler calling convention; a \
                         hand-written flag-taking entry needs an explicit audited contract.",
                        definition.label(),
                        discovered.evidence,
                        kept.evidence
                    ),
                    ..routine.clone()
                },
            );
        }
        Ok(contracts)
    }
}
